src/Microsoft.Extensions.Logging.Abstractions/Internal/FormattedLogValues.cs

```diff
@@ -14,17 +15,36 @@ namespace Microsoft.Extensions.Logging.Internal
     /// </summary>
     public class FormattedLogValues : IReadOnlyList<KeyValuePair<string, object>>
     {
+        internal const int MaxCachedFormatters = 1024;
         private const string NullFormat = "[null]";
+        private static int _count;
         private static ConcurrentDictionary<string, LogValuesFormatter> _formatters = new ConcurrentDictionary<string, LogValuesFormatter>();
         private readonly LogValuesFormatter _formatter;
         private readonly object[] _values;
         private readonly string _originalMessage;
 
+        // for testing purposes
+        internal LogValuesFormatter Formatter => _formatter;
+
         public FormattedLogValues(string format, params object[] values)
         {
             if (values?.Length != 0 && format != null)
             {
-                _formatter = _formatters.GetOrAdd(format, f => new LogValuesFormatter(f));
+                if (_count >= MaxCachedFormatters)
+                {
+                    if (!_formatters.TryGetValue(format, out _formatter))
+                    {
+                        _formatter = new LogValuesFormatter(format);
+                    }
+                }
+                else
+                {
+                    _formatter = _formatters.GetOrAdd(format, f =>
+                    {
+                        Interlocked.Increment(ref _count);
```

Comment:
I don't think 100% correctness is required here, we are preferring performance over it.